src/integrator_interface.jl

```diff
@@ -37,34 +37,34 @@ end
 Calculate next step of `integrator`.
 """
 function perform_step!(integrator::DDEIntegrator)
-    # update previous time to extrapolate from current interval
-    integrator.tprev = integrator.t
-
     # update ODE integrator
     integrator.integrator.uprev = integrator.uprev
-    integrator.integrator.tprev = integrator.tprev
-    integrator.integrator.fsalfirst = integrator.fsalfirst
+    integrator.integrator.tprev = integrator.t # to extrapolate from current interval
     integrator.integrator.t = integrator.t
     integrator.integrator.dt = integrator.dt
 
-    # if dt is greater than the minimal lag, then it's explicit so use fixed-point iteration
-    if integrator.dt > minimum(integrator.prob.lags)
+    # separate u to avoid influence of calculations on interpolation and to calculate
+    # residuals
+    if typeof(integrator.u) <: AbstractArray
+        recursivecopy!(integrator.u_cache, integrator.u)
+    else
+        integrator.u_cache = integrator.u
+    end
+    integrator.integrator.u = integrator.u_cache
 
-        # save these values to correct the extrapolation after the last iteration
-        tprev_cache = integrator.tprev
-        t_cache = integrator.t # same as tprev_cache?
-        uprev_cache = integrator.uprev
+    # separate k to avoid influence of calculations on interpolation
+    # constant caches that contain fsalfirst must be updated for interpolation
+    # has to be changed all such caches - move to OrdinaryDiffEq?
+    if typeof(integrator.cache) <: Union{BS3ConstantCache}
```

Comment:
That's exactly one of the reasons why I was confused about this part. So the problem in our case is rather that for in-place algorithms with `k[1] = integrator.fsalfirst` and `k[2] = integrator.fsallast` (like BS3Cache) we always automatically have `k[1] == k[2]` at this point.... Maybe it's best to always separate `k` by `integrator.integrator.k = recursivecopy(integrator.k)` right after initialization and then use `recursivecopy!` in all other places? But I guess I have to try it to see whether it actually works...